File: scrapers/armstrong.rb
require "nokogiri"
require "open-uri"

require "./models/models.rb"

ROOT_URL = "https://www.armstrongflooring.com"
PRODUCTS_URL = "https://www.armstrongflooring.com/commercial/en-us/products.html"

MANUFACTURER = "Armstrong"
TYPE = "Flooring"

products_doc = Nokogiri::HTML(open(PRODUCTS_URL))
count = 0

products_doc.css(".card__info h3 a").each do |link|
  collection_name = link.children.last.text.strip
  collection_url = ROOT_URL + link['href']

  collection_doc = Nokogiri::HTML(open(collection_url))
  collection_doc.css(".card--collection h4 a").each do |link|
    product_line_name = link.content.strip
    product_line_url = ROOT_URL + link["href"]

    product_line_doc = Nokogiri::HTML(open(product_line_url))
    product_line_doc.css(".card__info").each do |c|
      link = c.children[1]
      product_name = link.text.strip
      product_url = ROOT_URL + link["href"]

      product_doc = Nokogiri::HTML(open(product_url))
      product_info_table = product_doc.css(".item-differentiators").first
      product_infos = product_info_table.css("tr td")

      model_num = product_infos[0].text.strip
      color = product_infos[2].text.strip
      size = product_infos[3].text.strip

      Finishes::Option.create({
        "Name": product_name,
        "Type": TYPE,
        "URL": product_url,
        "Manufacturer": MANUFACTURER,
        "Manufacturer Subtype": collection_name,
        "Manufacturer Product Line": product_line_name,
        "Manufacturer Model #": model_num,
        "Info": [
          "Color: #{color}",
          "Size: #{size.gsub(/\s{2,}/, "\n")}",
        ].join("\n"),
        "User": [Finishes::Option::FINISH_VISION_VR_TEAM_USER_ID]
      })

      sleep 0.5

      count = count + 1
    end
  end
end

puts "Added #{count} finishes from #{MANUFACTURER}."

File: scrapers/armstrong.js
// Go to their listings page, load all of the products you want to scrape.  Run the script in the console.
//
var parser = new DOMParser();
var header = document.getElementsByTagName("h1")[0];
var labels = document.getElementsByClassName("card__info");
var sum = "";
var copy_ref = copy;

var type = "Flooring";
var manufacturer = "Armstrong";
var manufacturerSubtype = header.children[0].childNodes[2].textContent.trim();
var manufacturerProductLine = header.childNodes[header.childNodes.length-1].textContent.trim();

console.log(manufacturerSubtype);
console.log(manufacturerProductLine);

function getElementByXpath(path, doc) {
  return document.evaluate(path, doc, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue;
}

function scrapeURL(index) {
  if (index >= labels.length) {
    console.log(sum);
    copy_ref(sum);
    return;
  }

  console.log(index);

  var $link = labels[index].children[0];
  var name = $link.textContent.trim();
  var url = $link.href;

  fetch(url).then(function (res) {
    res.text().then(function (text) {
      var doc = parser.parseFromString(text, "text/html");
      console.log(doc);

      var model_num = getElementByXpath("//table[contains(@class, \"item-differentiators\")]/tbody[1]/tr[1]/td[1]", doc).textContent.trim();
      var description = [
        "Collection: " + getElementByXpath("//table[contains(@class, \"item-differentiators\")]/tbody[1]/tr[2]/td[1]", doc).textContent.trim(),
        "Design: " + getElementByXpath("//table[contains(@class, \"item-differentiators\")]/tbody[1]/tr[5]/td[1]", doc).textContent.trim(),
        "Color: " + getElementByXpath("//table[contains(@class, \"item-differentiators\")]/tbody[1]/tr[3]/td[1]", doc).textContent.trim(),
        "Size: " + getElementByXpath("//table[contains(@class, \"item-differentiators\")]/tbody[1]/tr[4]/td[1]", doc).textContent.trim().split("\n")[0].trim(),
      ].join("|");

      sum += [name, type, url, manufacturer, manufacturerSubtype, manufacturerProductLine, model_num, description].join("\t") + "\n";

      scrapeURL(index + 1);
    });
  });
}

scrapeURL(0);
